Add KS tests for weighted sampling

dhardy opened the pull request:
- [ ] Added a `CHANGELOG.md` entry

# Motivation

Some of these are non-trivial distributions we didn't really test before.

To validate solution of #1476.

# Details

Single-element weighted sampling is simple enough.

`fn choose_two_iterator` is also simple enough: there are no weights, so we can just assign each pair of results a unique index in the list of 100 * 99 / 2 possibilities (nothing that we sort pairs since the order of chosen elements is not specified).

`fn choose_two_weighted_indexed` gets a bit more complicated; I choose to approach it by building a table for the CDF of size `num*num` including impossible variants. Most of the tests don't pass, so there must be a mistake here.

Aside: using `let key = rng.random::<f64>().ln() / weight;` (`src/seq/index.rs:392`) may help with #1476 but does not fix the above.

dhardy commented:
I can confirm that `choose_multiple_weighted` has a significant problem, since sampling two elements from `0, 1, 2` with weights `1, 1/2, 1/3` a million times and sorting yields 532298 counts of `(0, 1)`, 338524 counts of `(0, 2)` and 129178 counts of `(1, 2)`. (Unlike #1476, this example does not require very small weights.)

dhardy commented:
I fixed my calculation of the CDF, found a variant which failed like #1476, fixed this by taking the logarithm of keys, and applied some optimisation to the Efraimidis-Spirakis algorithm.

benjamin-lieser commented on src/seq/index.rs:
```diff
@@ -387,37 +387,49 @@ where
+        if weight > 0.0 {
+            x -= weight;
+            if x <= 0.0 {
+                let t = core::f64::consts::E.powf(candidates[0].key * weight);
```
Is there a reason not to use `exp`? I guess it will be faster in general

benjamin-lieser commented on src/seq/index.rs:
```diff
@@ -387,37 +387,49 @@ where
-    // the *smallest* keys at the beginning of the list in `O(n)` time, which
-    // provides equivalent information about the elements with the *greatest* keys.
-    let (_, mid, greater) = candidates.select_nth_unstable(avail - amount.as_usize());
+    let mut x = rng.random::<f64>().ln() / candidates[0].key;
```
Maybe name `T` to be close to the original paper?

benjamin-lieser replied:
Sry, I was misunderstanding. Maybe mention that is implements A-ExpJ and not the first one in the paper

benjamin-lieser commented on src/seq/index.rs:
```diff
@@ -387,37 +387,49 @@ where
+                let t = core::f64::consts::E.powf(candidates[0].key * weight);
+                let key = rng.random_range(t..1.0).ln() / weight;
+                candidates[0] = Element { index, key };
+                candidates.sort_unstable();
```
I guess it is very likely that a tree data structure would perform much faster if amount is big enough. Not sure where the threshold is. Depending on `sort_unstable` is could even perform particularly worse on an almost sorted slice.

dhardy replied:
Good point, though I won't address it now.

benjamin-lieser replied:
https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html
should be always faster as it also stores its elements in a `Vec` should be an easy change.

benjamin-lieser reviewed (commented):
Looks correct. We might want to test how the performance is for big `amount`

benjamin-lieser commented on distr_test/tests/weighted.rs:
```diff
@@ -0,0 +1,234 @@
+}
+
+#[test]
+fn choose_two_weighted_indexed() {
```
I mean using KS for these distributions (chi squared would be more straight forward), Adapter I think is fine.

dhardy commented:
@benjamin-lieser I addressed most of your feedback, using `BinaryHeap` as suggested. Best-of-three runs before this change (a039a7fdf9):
```
running 3 tests
test choose_weighted_indexed ... ok
test choose_one_weighted_indexed ... ok
test choose_two_weighted_indexed ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out; finished in 3.69s
```
Best-of-three after (831353ff55):
```
running 3 tests
test choose_weighted_indexed ... ok
test choose_one_weighted_indexed ... ok
test choose_two_weighted_indexed ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out; finished in 3.98s
```
Yes, that's a regression, though given that we're using `amount ∊ {1, 2}` these tests aren't very representative benchmarks. I agree that in theory this approach should scale better to large `amount`s.

benjamin-lieser reviewed (approved):
I agree that we should use this version. The other one might be faster in our benchmarks where length and amount are not too large, but it is not true to the idea of the algorithm and users might expect it to have the correct memory footprint.
